src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+
+
+
+(** {2 Fundamentals} *)
+
+(** {3 Promises} *)
 
 type +'a t
-  (** The type of threads returning a result of type ['a]. *)
+(** Promises for values of type ['a].
+
+    A {b promise} is a memory cell that is always in one of three {b states}:
+
+    - {e fulfilled}, and containing one value of type ['a],
+    - {e rejected}, and containing one exception, or
+    - {e pending}, in which case it may become fulfilled or rejected later.
+
+    A {e resolved} promise is one that is either fulfilled or rejected, i.e. not
+    pending. Once a promise is resolved, its content cannot change. So, promises
+    are {e write-once references}. The only possible state changes are (1) from
+    pending to fulfilled and (2) from pending to rejected.
+
+    Promises are typically “read” by attaching {b callbacks} to them. The most
+    basic functions for that are {!Lwt.bind}, which attaches a callback that is
+    called when a promise becomes fulfilled, and {!Lwt.catch}, for rejection.
+
+    Promises variables of this type, ['a Lwt.t], are actually {b read-only} in
+    Lwt. Separate {e resolvers} of type ['a ]{!Lwt.u} are used to write to them.
+    Promises and their resolvers are created together by calling {!Lwt.task}.
+    There is one exception to this: most promises can be {e canceled} by calling
+    {!Lwt.cancel}, without going through a resolver. *)
+
+type 'a u
+(** Resolvers for promises of type ['a ]{!Lwt.t}.
+
+    Each resolver can be thought of as the {b write end} of one promise. It can
+    be passed to {!Lwt.wakeup_later}, {!Lwt.wakeup_later_exn}, or
+    {!Lwt.wakeup_later_result} to resolve that promise. *)
+
+val task : unit -> ('a t * 'a u)
+(** Creates a new pending {{: #TYPEt} promise}, paired with its {{: #TYPEu}
+    resolver}.
+
+    It is rare to use this function directly. Many helpers in Lwt, and Lwt-aware
+    libraries, call it internally, and return only the promise. You then chain
+    the promises together using {!Lwt.bind}.
+
+    However, it is important to understand [Lwt.task] as the fundamental promise
+    “constructor.” All other functions that evaluate to a promise can be, or
+    are, eventually implemented in terms of it. *)
+
+
+
+(** {3 Resolving} *)
+
+val wakeup_later : 'a u -> 'a -> unit
+(** [Lwt.wakeup_later r v] {e fulfills}, with value [v], the {e pending}
+    {{: #TYPEt} promise} associated with {{: #TYPEu} resolver} [r]. This
+    triggers callbacks attached to the promise.
+
+    If the promise is not pending, [Lwt.wakeup_later] raises
+    {{: https://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html#VALinvalid_arg}
+    [Pervasives.Invalid_argument]}, unless the promise is {{: #VALcancel}
+    canceled}. *)
+
+val wakeup_later_exn : _ u -> exn -> unit
+(** [Lwt.wakeup_later_exn r exn] is like {!Lwt.wakeup_later}, except it
+    {e rejects} the associated {e pending} {{: #TYPEt} promise} with [exn]
+    instead. *)
 
```

Comment:
The place where it's mentioned earlier is actually a hyperlink, so the user can find it easily from there. Also, I think we should start getting away from `Lwt.result` (not `Lwt_result`!), and towards `Lwt_result` :p

Reply:
That said, if you or others use `Lwt.result`, `wakeup_later_result`, etc., a lot, we can move the function up. I'd rather not link to it from directly inside the description of `wakeup_later_exn`, because then it seems we'd also have to link to it from `wakeup_later`, and it becomes weird to give it such special status eventually.